### PR TITLE
server: Add TTL checker

### PR DESCRIPTION
**Pull request by Connor1996**
### What problem does this PR solve?

part of https://github.com/tikv/tikv/pull/9595

### What is changed and how it works?

- Add ttl checker to compact files of invalidated 
- Add grafana metrics

image

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 
- None

### CODE REVIEW

File: src/server/ttl/ttl_checker.rs [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        thread::sleep(Duration::from_secs(40));
+        TTL_CHECKER_PROCESSED_REGIONS_GAUGE.set(0);
+    }
+
+    fn get_interval(&self) -> Duration {
+        self.poll_interval
+    }
+}
+
+fn check_ttl_for_range<E: KvEngine>(
```
Comment (Little-Wallace) [inline src/server/ttl/ttl_checker.rs:141]: Shall it be named to `check_ttl_and_compact_range` ?

File: src/server/ttl/ttl_checker.rs [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+            TTL_CHECKER_ACTIONS_COUNTER_VEC
+                .with_label_values(&["error"])
+                .inc();
+            continue;
+        }
+        compact_range_timer.observe_duration();
+        TTL_CHECKER_ACTIONS_COUNTER_VEC
+            .with_label_values(&["compact"])
+            .inc();
+        thread::sleep(Duration::from_secs(2));
```
Comment (Little-Wallace) [inline src/server/ttl/ttl_checker.rs:202]: Do not use magic number.

File: src/server/ttl/ttl_checker.rs [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+    }
+
+    info!(
```
Comment (Little-Wallace) [inline src/server/ttl/ttl_checker.rs:205]: The number of log  record may be too large when there are too many region need to process every round

**Review by Little-Wallace — APPROVED** [review]
LGTM

File: src/import/sst_service.rs [hunk 4]
```diff
@@ -396,21 +396,6 @@ where
                     "output_level" => ?output_level,
                 ),
             }
-            let res = engine.compact_files_in_range(start, end, output_level);
```
Comment (BusyJay) [inline src/import/sst_service.rs:399]: Why delete this?

File: cmd/src/server.rs [hunk 5]
```diff
@@ -497,14 +489,20 @@ impl<ER: RaftEngine> TiKVServer<ER> {
...
-            RaftRouter<RocksEngine, ER>,
-        >,
-    ) -> Arc<ServerConfig> {
+    fn init_servers(&mut self) -> Arc<ServerConfig> {
+        let gc_worker = self.init_gc_worker();
+        let mut ttl_checker = Box::new(LazyWorker::new("ttl-checker"));
+        let ttl_scheduler = ttl_checker.scheduler();
+
         let cfg_controller = self.cfg_controller.as_mut().unwrap();
+        cfg_controller.register(
```
Comment (BusyJay) [inline cmd/src/server.rs:498]: Not sure if it's OK to adjust the order. /cc @NingLin-P

File: components/engine_rocks/src/compact.rs [hunk 6]
```diff
@@ -92,14 +86,53 @@ impl CompactExt for RocksEngine {
...
+        )
+    }
+
+    fn compact_files_cf(
+        &self,
+        cf: &str,
+        mut files: Vec<String>,
+        output_level: Option<i32>,
+        max_subcompactions: u32,
+        without_l0: bool,
```
Comment (BusyJay) [inline components/engine_rocks/src/compact.rs:104]: s/without_l0/exclude_l0/

File: components/engine_rocks/src/compact.rs [hunk 7]
```diff
@@ -92,14 +86,53 @@ impl CompactExt for RocksEngine {
...
+            .unwrap_or(DBCompressionType::No);
+        let output_file_size_limit = cf_opts.get_target_file_size_base() as usize;
+
+        if without_l0 {
+            let cf_meta = db.get_column_family_meta_data(handle);
+            let mut l0_files = vec![];
+            for f in cf_meta.get_levels()[0].get_files() {
+                l0_files.push(f.get_name());
+            }
+            files.retain(|f| !l0_files.iter().any(|n| f.ends_with(n)));
```
Comment (BusyJay) [inline components/engine_rocks/src/compact.rs:123]: Why not just use `cf_meta.get_levels()[0].get_files()` directly?

File: src/storage/config.rs [hunk 8]
```diff
@@ -135,6 +143,11 @@ impl ConfigManager for StorageConfigManger {
                         .set(size.0 as f64);
                 }
             }
+        } else if let Some(v) = change.remove("ttl_check_poll_interval") {
```
Comment (BusyJay) [inline src/storage/config.rs:146]: Any test case?

File: src/import/sst_service.rs (hunk 4 above)
Comment (Little-Wallace) [reply src/import/sst_service.rs:399]: It is repeated code.....

Comment (Connor1996) [reply src/import/sst_service.rs:399]: yes, it is duplicated.

File: src/storage/config.rs (hunk 8 above)
Comment (Little-Wallace) [reply src/storage/config.rs:146]: See test case in src/config.rs

Comment (Connor1996) [reply src/storage/config.rs:146]: covered by `test_change_ttl_check_poll_interval`, and I tested it manually too
image

File: cmd/src/server.rs (hunk 5 above)
Comment (NingLin-P) [reply cmd/src/server.rs:498]: It is OK, `ConfigController` use `Arc<RwLock<_>>` internally.

**Review by Little-Wallace — APPROVED** [review]
LGTM

File: src/server/ttl/ttl_checker.rs [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+                }),
+            ) {
+                error!(?e; "ttl checker: failed to get next region information");
+                TTL_CHECKER_ACTIONS_COUNTER_VEC
+                    .with_label_values(&["error"])
+                    .inc();
+                continue;
+            }
+
+            match rx.recv() {
```
Comment (BusyJay) [inline src/server/ttl/ttl_checker.rs:105]: This will block the whole thread.

File: src/server/ttl/ttl_checker.rs [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+                Ok(None) => {}
+                Ok(Some((start_key, end_key))) => {
+                    let start = keys::data_key(&start_key);
+                    let end = keys::data_end_key(&end_key);
+                    check_ttl_and_compact_files(&self.engine, &start, &end, true);
+                    if !end_key.is_empty() {
```
Comment (BusyJay) [inline src/server/ttl/ttl_checker.rs:111]: Some keys can be leak if range `["key", "")` was in the node and then got removed.

File: cmd/src/server.rs [hunk 11]
```diff
@@ -719,6 +717,12 @@ impl<ER: RaftEngine> TiKVServer<ER> {
         }
 
         initial_metric(&self.config.metric);
+        ttl_checker.start_with_timer(TTLChecker::new(
```
Comment (BusyJay) [inline cmd/src/server.rs:720]: If ttl is not enabled, no reason to do a full check every 24h.

File: src/server/ttl/ttl_checker.rs (hunk 9 above)
Comment (Connor1996) [reply src/server/ttl/ttl_checker.rs:105]: we do want to do it one by one

File: src/server/ttl/ttl_checker.rs (hunk 10 above)
Comment (Connor1996) [reply src/server/ttl/ttl_checker.rs:111]: It doesn't matter, they will be checked in the next round.

Comment (Connor1996) [reply src/server/ttl/ttl_checker.rs:111]: oh, I got want you mean

File: src/server/ttl/ttl_checker.rs (hunk 9 above)
Comment (BusyJay) [reply src/server/ttl/ttl_checker.rs:105]: Using async/await won't change the order.

Comment (Connor1996) [reply src/server/ttl/ttl_checker.rs:105]: it's in a single thread not sharing with others.

**Review by BusyJay — APPROVED** [review]

**Pull request merged**